### PR TITLE
AbstractPoint, Point, and WitnessSet

### PR DESCRIPTION
**Pull request by antonleykin**
The main changes implemented are those related to `Point` and `WitnessSet` --- both types inherit from (immutable) HashTables now.

`Point` is now a derived from `AbstractPoint`.
The idea behind the abstract type `AbstractPoint` is to have other (more efficient? specialized?) implementations of similar to `Point`.
(`AbstractPoint` establishes "abstract" methods, e.g., `coordinates, that the types that inherit from it need to define.)

The authors of the packages, please make sure if the changes I made are appropriate.
It also is recommended to see if `Point` can be replaced with `AbstractPoint` in implementations of your methods ---
this would make your package compatible with the future types of "points".

I ran (successfully) `check` and `installPackage` on
* NAGtypes
* NumericalAlgebraicGeometry
* PHCpack @janverschelde 
* Bertini @JoseMath 
* MonodromySolver @timduff35 
* NumericalImplicitization @jchen419 
* NumericalCertification @klee669 
* NoetherianOperators @haerski 
* NumericalSchubertCalculus
* SemidefiniteProgramming @diegcif

Note: new versions of `*/examples` for `PHCpack`, `Bertini`, and `NumericalSchubertCalculus` were not committed.

### CODE REVIEW

File: M2/Macaulay2/packages/NAGtypes/Point-abstract.m2 [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	if na > 1 or nb > 1 then norm(2,(a-b)) < o.Tolerance * max(na,nb)
+	else norm(2,a-b) < o.Tolerance -- in case both points are close to the origin, absolute error is looked at 
+	)
+    )
+-*
+areEqual (AbstractPoint,BasicList) := o -> (a,b) -> areEqual(coordinates a, toList b)
+areEqual (BasicList,AbstractPoint) := o -> (a,b) -> areEqual(toList a, coordinates b)
+*-
+
+isGEQ = method(TypicalValue=>Boolean, Options=>{Tolerance=>1e-6})
```
Comment (DanGrayson) [inline M2/Macaulay2/packages/NAGtypes/Point-abstract.m2:66]: Is there an application for this ordering relation on points?

**Comment by mikestillman** [comment]
@antonleykin It seems no other authors have commented here on your changes.  Have you received comments offline?  Can you respond to @DanGrayson 's comments in this conversation?  After that, I say we accept this pull request.  @antonleykin Did yo want me to look at some parts of this more carefully for suggestions, or is it ready to go?

**Comment by diegcif** [comment]
I approve the changes in SemidefiniteProgramming

**Comment by DanGrayson** [comment]
> @janverschelde @JoseMath @timduff35 @jchen419 @klee669 @haerski @diegcif -- do you have any comments about the changes to your packages?

**Comment by jchen419** [comment]
The changes in ``NumericalImplicitization`` look fine to me.

File: M2/Macaulay2/packages/NAGtypes/Point-abstract.m2 [hunk 2]
```diff
@@ -0,0 +1,78 @@
+-- Abstract classes ------------------------
+AbstractPoint = new Type of HashTable
+
+AbstractPoint.synonym = "point"
+texMath AbstractPoint := x -> texMath coordinates x
+net AbstractPoint := p -> net coordinates p
+
+AbstractPoint == AbstractPoint := (a,b) -> areEqual(a,b) -- the default Tolerance is used
+AbstractPoint ? AbstractPoint := (a,b) -> if isGEQ(a,b) then symbol > else symbol <
```
[the remark this answers is not in the document]
Comment (antonleykin) [reply M2/Macaulay2/packages/NAGtypes/Point-abstract.m2:9]: Yes, but it is not useful for a typical application, which is sorting.

File: M2/Macaulay2/packages/NAGtypes/Point-abstract.m2 (hunk 1 above)
Comment (antonleykin) [reply M2/Macaulay2/packages/NAGtypes/Point-abstract.m2:66]: Sorting. If one wants to compare lists of solutions returned by homotopy continuation routines they need to be sorted first.

**Comment by antonleykin** [comment]
> @antonleykin It seems no other authors have commented here on your changes. Have you received comments offline? 

Not many. Most of the type "it is probably fine".

> Can you respond to @DanGrayson 's comments in this conversation? 

Dan's comments went to "Promotions" on gmail :)

> After that, I say we accept this pull request. @antonleykin Did yo want me to look at some parts of this more carefully for suggestions, or is it ready to go?

From my point of view, it's ready to go.

**Comment by JoseMath** [comment]
I also approve the changes :)

**Comment by haerski** [comment]
`NoetherianOperators` looks good!

**Pull request merged**